core/controllers/reader.py

```diff
@@ -601,12 +601,14 @@ def post(self, exploration_id):
 
         if user_id and story_id:
             story = story_fetchers.get_story_by_id(story_id)
-            if story:
+            if story is not None:
                 learner_progress_services.mark_story_as_incomplete(
                     user_id, story.id)
-            if story.corresponding_topic_id:
-                learner_progress_services.mark_topic_as_partially_learnt(
-                    user_id, story.corresponding_topic_id)
+                if story.corresponding_topic_id is not None:
```

Comment:
Suppose the topic is already learnt (because some other canonical story in it has been completed). Wouldn't marking it as partially learnt here be incorrect?

Also please note that the "learnt" criterion applies only to canonical stories, not all stories. Please make sure, throughout your PR, that you're checking canonical stories only (and not stories in the "additional_story_ids" field).

Reply:
Before marking the topic as partially learnt, the mark_topic_as_partially_learnt will check if the topic is already marked as learnt or not. Yeah, throughout the PR we are checking only the canonical stories.